cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,619 @@
+     -- changed.
+     --
+   | MonitorFileHashed !FilePath
+
+     -- | Monitor a single non-existent file for changes. The monitored file
+     -- is considered to have changed if it exists.
+     --
+   | MonitorNonExistentFile !FilePath
+
+     -- | Monitor a set of files identified by a file glob. The monitored glob
+     -- is considered to have changed if the set of files matching the glob
+     -- changes (i.e. creations or deletions), or if the modification time and
+     -- content hash of any matching file has changed.
+     --
+   | MonitorFileGlob !FilePathGlob
+
+  deriving (Show, Generic)
+
+instance Binary MonitorFilePath
+
+-- | A file path specified by globbing
+--
+data FilePathGlob
+   = GlobDir  !Glob !FilePathGlob
+   | GlobFile !Glob
+  deriving (Eq, Show, Generic)
+
+instance Binary FilePathGlob
+
+
+monitorFileSearchPath :: [FilePath] -> FilePath -> [MonitorFilePath]
+monitorFileSearchPath notFoundAtPaths foundAtPath =
+    MonitorFile foundAtPath
+  : map MonitorNonExistentFile notFoundAtPaths
+
+monitorFileHashedSearchPath :: [FilePath] -> FilePath -> [MonitorFilePath]
+monitorFileHashedSearchPath notFoundAtPaths foundAtPath =
+    MonitorFileHashed foundAtPath
+  : map MonitorNonExistentFile notFoundAtPaths
+
+-- | A file used to store the state of a file monitor.
+--
+-- It is typed to ensure it's used at consistent types.
+--
+newtype FileMonitorCacheFile a b
+      = FileMonitorCacheFile { monitorCacheFilePath :: FilePath }
+  deriving Show
+
+------------------------------------------------------------------------------
+-- Implementation types, files status
+--
+
+data MonitorStateFileSet
+   = MonitorStateFileSet !(Map FilePath MonitorStateFile)
+                         ![FileGlobMonitorState]
+  deriving Show
+
+--instance Monoid MonitorStateFileSet where
+--  mempty = MonitorStateFileSet Map.empty []
+--  MonitorStateFileSet a b `mappend` MonitorStateFileSet x y =
+--    MonitorStateFileSet (a<>x) (b<>y)
+
+type Hash = Int
+type ModTime = UTCTime
+
+-- | The state necessary to determine whether monitored files have changed.
+--
+-- This covers all the cases of 'MonitorFilePath' except for globs which is
+-- covered separately by 'FileGlobMonitorState'.
+--
+data MonitorStateFile
+   = MonitorStateFile       !ModTime         -- ^ cached file mtime
+   | MonitorStateFileHashed !ModTime !Hash   -- ^ cached file mtime and content hash
+   | MonitorStateFileNonExistent
+
+     -- | These two are to deal with the situation where we've been asked
+     -- to monitor a file that's expected to exist, but when we come to
+     -- check it's status, it no longer exists.
+   | MonitorStateFileChanged
+   | MonitorStateFileHashChanged
+  deriving (Show, Generic)
+
+instance Binary MonitorStateFile
+
+-- | The state necessary to determine whether the files matched by a globbing
+-- match have changed.
+--
+data FileGlobMonitorState
+   = MonitorStateGlobDirs  !Glob !FilePathGlob
```

Comment:
Yeah, this stuff is subtle. The big picture is we're efficiently determining if a set of files has changed or not. But rather than just a set of files, we can also monitor files defined by a file glob. Doing that efficiently is a little more tricky.